flextape/service/queue.go

```diff
@@ -0,0 +1,137 @@
+package service
+
+import (
+	"sort"
+)
+
+// invocationq is a simple queue of invocations.
+//
+// Beside implementing the classic Enqueue and Dequeue methods:
+// - It partially implements the sort.Interface to allow the easy
+//   implementation of invocation prioritization policies.
+// - It maintains per-entry state (QueueID) so it's easy to tell
+//   the position of each invocation by just looking at that value.
+type invocationq []*invocation
+
+func (iq *invocationq) Enqueue(x *invocation) Position {
+	position := Position(len(*iq))
+	offset := QueueID(1)
+	if len(*iq) > 0 {
+		offset = (*iq)[0].QueueID
+	}
+
+	x.QueueID = offset + QueueID(position)
+	(*iq) = append(*iq, x)
```

Comment:
Same question here about locking

Reply:
Similar answer to above. The lock ordering fears are worse here as `Walk` and `Filter` take callbacks, it would be easy to introduce a deadlock accidentally. Eg, add more risks than benefits?